### ISSUE TITLE
Docker Image Removal Bug By Id

### ISSUE THREAD

**Issue by danielrolnick** [docker/docker#11414]
There is a bug with image removal where an image with multiple tags will be allowed to be removed regardless of whether that image is in use or not if you try to remove by id.

Steps to Reproduce:

    # docker pull drolnick/spin:v2
    Pulling repository drolnick/spin
    38b8f2f5dc2b: Download complete 
    511136ea3c5a: Download complete 
    36669626e49c: Download complete 
    d5570ef1464a: Download complete 
    b3936f269611: Download complete 
    Status: Downloaded newer image for drolnick/spin:v2

    # docker run -d drolnick/spin:v2
    800851c61645b15161e3631dac59db6fa2dc98de9d43e3ac17c688ef7b7193d6

    # docker rmi 38b8f2f5dc2b
    Untagged: drolnick/spin:v2
    Untagged: drolnick/spin:v2_1
    Untagged: drolnick/spin:v2_3
    Deleted: 38b8f2f5dc2b34f428f778abca5c982ec8b36d0e4c63d3e943882eea2dac1946
    Deleted: b3936f26961179f0fe7e217c13236497af012980139c307d6402d9d4d6c5d50a
    Deleted: d5570ef1464a43fe282dd2705b38a2d739812b0b8036a49cfa09811737cfbed4
    Deleted: 36669626e49c623f3b2aea41052bbc9de0506807d92137bf76c0fd3264eaff34
    Deleted: 511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158

    # docker images
    REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE

Compare that to the effect that happens when you do the following

    # docker pull drolnick/spin:v2
    Pulling repository drolnick/spin
    38b8f2f5dc2b: Download complete 
    511136ea3c5a: Download complete 
    36669626e49c: Download complete 
    d5570ef1464a: Download complete 
    b3936f269611: Download complete 
    Status: Downloaded newer image for drolnick/spin:v2

    # docker run -d drolnick/spin:v2
    800851c61645b15161e3631dac59db6fa2dc98de9d43e3ac17c688ef7b7193d6

    # docker rmi drolnick/spin:v2
    Untagged: drolnick/spin:v2

    # docker rmi drolnick/spin:v2_1
    Untagged: drolnick/spin:v2_1

    # docker rmi drolnick/spin:v2_3
    Error response from daemon: Conflict, cannot delete 38b8f2f5dc2b because the running container 800851c61645 is using it, stop it and use -f to force
    FATA[0000] Error: failed to remove one or more images 

    # docker images
    REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
    drolnick/spin       v2_3                38b8f2f5dc2b        About an hour ago   84.98 MB

We tracked this down to **docker/daemon/image_delete.go**  The function DeleteImage only checks  canDeleteImage if there is only a single tag mapping to an image.  

    # docker version
    Client version: 1.5.0
    Client API version: 1.17
    Go version (client): go1.4.1
    Git commit (client): a8a31ef
    OS/Arch (client): linux/ad64
    Server version: 1.5.0
    Server API version: 1.17
    Go version (server): go1.4.1
    Git commit (server): a8a31ef

**Comment by colinmorelli** [comment]
In the most common case (that @danielrolnick mentioned above), the implications of this bug are that the docker daemon loses track of the images. The daemon believes the images have been deleted, but all layers/diffs in the aufs directory are still present.

However, we have seen (in specific scenarios that we are trying to replicate) that this has actually caused  corruption of the AUFS filesystem that results in our containers failing (the mounts report stale file handles). When investigating `/sys/fs/aufs`, we see a tree that looks like:

```
/var/lib/docker/aufs/diff/9b0f8af8b250f69018b9ad8be2a38f431f052c26070999a5a853117ef3434d33=rw
/var/lib/docker/aufs/diff/9b0f8af8b250f69018b9ad8be2a38f431f052c26070999a5a853117ef3434d33-init=ro+wh
/var/lib/docker/aufs/diff/0356cece37f331dbf58a7791d3e382c3a1a94925b2f78049c1a0532f7066c776=ro+wh
/var/lib/docker/aufs/diff/97ca33ff69ce3356de59eb8f51db07add123b0b1ea60afac9334d7d301b17571=ro+wh
/var/lib/docker/aufs/diff/ac9a4e035be92cb89ff42c22ae59258d9fbdc5f8b39ba51768fad99666322e69=ro+wh
/var/lib/docker/aufs/diff/f6fab3b798be3174f45aa1eb731f8182705555f89c9026d8c1ef230cbf8301dd=ro+wh
/var/lib/docker/aufs/diff/f10807909bc552de261ca7463effc467600c3dca68d8e7704425283a6911d2ca=ro+wh
/var/lib/docker/aufs/diff/511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158=ro+wh
/var/lib/docker/aufs/diff/6b8af1d0e4a5dbf4383dc77d35f4f8dc78aefbeb4ce309e6cbc0df081228e126=ro+wh
/var/lib/docker/aufs/diff/35907e19c7516b193cd2603eba74294108159156d9312342e7f34b11601d2f18=ro+wh
/var/lib/docker/aufs/diff/e931d87ef04c0628d7f3934ae61f5e5efb2e918a363854303ab04685a87b6132-removing\040(deleted)=ro+wh
/var/lib/docker/aufs/diff/3a827c23bfd56b9ad671d5c19ca8096e0ac3dff0278cd26f8d24108362a133a0=ro+wh
/var/lib/docker/aufs/diff/c667667f49b4c339b4b89ff65273bf43921c9e743231ccf87815a611fdb38847=ro+wh
/var/lib/docker/aufs/diff/07a6c5b3b160da1e648955e3236a530c7cf3483ef8018d1cba3739e3edfd77e8=ro+wh
/var/lib/docker/aufs/diff/fae229a5cb2207dede4f300169c72b47771da5eec7101f7bf428ef16284df3e5=ro+wh
/var/lib/docker/aufs/diff/aa74c8db2e1ca874f364995aefaea646e6dd91281bb194ddafc201a1166b2ba9=ro+wh
```

As mentioned, we're still trying to replicate this particular issue, but we're fairly confident it stems from the fact that the daemon allows for these images to be deleted when they shouldn't be. When this happens, we see this in our `aufs/mnt` directory:

```
root@host:~# ls -al /var/lib/docker/aufs/mnt/
ls: cannot access /var/lib/docker/aufs/mnt/9b0f8af8b250f69018b9ad8be2a38f431f052c26070999a5a853117ef3434d33: Stale file handle
total 64
drwxr-xr-x 185 root root 81920 Mar 16 13:49 .
drwxr-xr-x   5 root root    40 Jul 31  2014 ..
??????????   ? ?    ?        ?            ? 9b0f8af8b250f69018b9ad8be2a38f431f052c26070999a5a853117ef3434d33
```

This then causes issues when attempting to interact with the container:

```
root@host:~# docker exec -it ee6b516658d0377e83c60bed21a60a399f84518b160aa57cc2fd8f0fe0a06e04 /bin/bash
2015/03/16 18:45:58 docker-exec: failed to exec: close open file descriptors open /proc/self/fd: no such file or directory
root@host:~# docker restart ee6b5
Error response from daemon: Cannot restart container ee6b5: fork/exec /var/lib/docker/init/dockerinit-1.5.0: stale NFS file handle
FATA[0010] Error: failed to restart one or more containers 
```

**Comment by colinmorelli** [comment]
Appears to stem from the following section of code:

```
	//If delete by id, see if the id belong only to one repository
	if repoName == "" {
		for _, repoAndTag := range repos {
			parsedRepo, parsedTag := parsers.ParseRepositoryTag(repoAndTag)
			if repoName == "" || repoName == parsedRepo {
				repoName = parsedRepo
				if parsedTag != "" {
					tags = append(tags, parsedTag)
				}
			} else if repoName != parsedRepo && !force && first {
				// the id belongs to multiple repos, like base:latest and user:test,
				// in that case return conflict
				return fmt.Errorf("Conflict, cannot delete image %s because it is tagged in multiple repositories, use -f to force", name)
			}
		}
	} else {
		tags = append(tags, tag)
	}
```

If I'm reading this correctly, if you are deleting by ID instead of by tag, then docker reverse looks up the image ID to get an array of all matching repos, and parses the tags out of each one of those. Those tags are then all appended to the `tags` array. Docker bypasses the `canDeleteImage` check unless the length of `repos <= 1` (which isn't true, becuase this image has multiple tags).

It then goes and untags everything in the `tags` array (which is an array of all tags, as determined via the reverse lookup). The image is now left with 0 tags. The following check is then `if len(tags) == 0` (which is now true, because all tags have been removed), so Docker deletes the image even though it bypassed the `canDeleteImage` method (which would have returned false).

We are working on a pull request.

**Comment by thaJeztah** [comment]
@colinmorelli is this related to https://github.com/docker/docker/pull/11058?

**Comment by colinmorelli** [comment]
Yes and no. We're not doing force deletions, but it's the same general area of code causing problems (as far as I can tell). Interestingly, we're seeing the inverse of what that PR suggests (we're seeing that, without a force delete, deleting an image by ID deletes all tags *and* the underlying image)

**Comment by thaJeztah** [comment]
@colinmorelli yeah, it sounded *different*, but I thought the other PR would be worth mentioning before you started coding :)

**Comment by x1022as** [comment]
@colinmorelli here is the difference:
Assume this case, `repo:tag1` `repo:tag2` ... with same imageID, `docker rmi $imageID` will untag all and delete the image because they have same repo name, the tags do not matter.

In another case, `repo:tag`, `repo1:tag1`, `repo2:tag2` ..., `docker rmi $imageID` will just untag a random image because of the different repo names.

#11058 will untag all these repo&tags(no matter repo names are the same or different) and delete the image with `docker rmi -f $imageID`.

Hope I've made this clear

**Comment by colinmorelli** [comment]
@x1022as that clears it up a lot. In our case, all instances have the same repo name, which explains the difference in behavior. I agree with #11058 (`docker rmi -f <image>`) should remove *all* repos/tags and the underlying image

**Comment by jfrazelle** [comment]
this is fixed on master and has a regression test I added as well see https://github.com/docker/docker/pull/12654/files

**Comment by softboysxp** [comment]
I'm re-opening this since it does not seem to be fixed in either master or the latest release (1.6.2):

The following pulls down two tagged repos pointing to the same image (in this case, nginx 1.9), then start up a container, "docker rmi" (without -f) still manages to delete all images even though the container is still running. (doing rmi with only a single tag will result in the conflict error as expected)

Reproduced on both 1.6.2 release and head of master (1f22676 as of this report) on Arch and Amazon Linux.

# ./docker version
Client version: 1.7.0-dev
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 1f22676
OS/Arch (client): linux/amd64
Server version: 1.7.0-dev
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 1f22676
OS/Arch (server): linux/amd64
# ./docker pull nginx:latest
latest: Pulling from nginx

39bb80489af7: Pull complete
df2a0347c9d0: Pull complete
dc2e1697e33e: Pull complete
e21d523a1481: Pull complete
5ec936b59c11: Pull complete
3ec5f57e729c: Pull complete
1f1cfc8b4072: Pull complete
514f4db63e53: Pull complete
e2fde5e7e71f: Pull complete
8cac0c007422: Pull complete
72d73c46937a: Pull complete
a785ba7493fd: Already exists
nginx:latest: The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security.

Digest: sha256:88f8d82bc9bc20ff80992cdeeee1dd6d8799cd36797b3653c644943e90b3acdf
Status: Downloaded newer image for nginx:latest
# ./docker pull nginx:1.9
1.9: Pulling from nginx

39bb80489af7: Already exists
df2a0347c9d0: Already exists
dc2e1697e33e: Already exists
e21d523a1481: Already exists
5ec936b59c11: Already exists
3ec5f57e729c: Already exists
1f1cfc8b4072: Already exists
514f4db63e53: Already exists
e2fde5e7e71f: Already exists
8cac0c007422: Already exists
72d73c46937a: Already exists
a785ba7493fd: Already exists
nginx:1.9: The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security.

Digest: sha256:5a3a97160f60870cc345bfe1261a53d41bbdd1736f1dcf60471ee0c9675482e9
Status: Downloaded newer image for nginx:1.9
# ./docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
docker-dev          master              9f36957c5e0b        6 minutes ago       1.567 GB
nginx               latest              a785ba7493fd        28 hours ago        132.8 MB
nginx               1.9                 a785ba7493fd        28 hours ago        132.8 MB
ubuntu              14.04               07f8e8c5e660        4 weeks ago         188.3 MB
# ./docker run -d nginx
5f4242636bc9f45614c18b5f3683f6b27ffe38a9de67bc338df2238cdaefbc77
# ./docker ps
CONTAINER ID        IMAGE               COMMAND                CREATED             STATUS              PORTS               NAMES
5f4242636bc9        nginx               "nginx -g 'daemon of   2 seconds ago       Up 1 seconds                            loving_albattani
# ./docker rmi a785ba7493fd
Untagged: nginx:1.9
Untagged: nginx:latest
Deleted: a785ba7493fd9009be71023045790697f334486b11a45f7e830652d39665a451
Deleted: 72d73c46937aaf8ca1efe566cac0564a05a0f1aeaef770658c8201726cfff50b
Deleted: 8cac0c007422e6c6cec5d15ad1dc84afdeabc09810dbacfbf5d585656430e29c
Deleted: e2fde5e7e71fc74acebe1d75c9e2131bc66b7453e06f1220cdaa1395f0cb0608
Deleted: 514f4db63e53ab02f33d811973fec63baf8018883b7afcaffc76a7f49540c3cd
Deleted: 1f1cfc8b4072488196bdf2c7638a50940849cca3d35f1e35141e82edb0a41789
Deleted: 3ec5f57e729cb2f888889bfe90498a49370cf68461ea6913f0cda191f629d738
Deleted: 5ec936b59c11fd943372e2f662baf6e385fdc909b8bbb1dd4750def52123352c
Deleted: e21d523a1481c4fccb7efb0ff9ff49e6b903ff63dc6fef977fa5970beefce0ea
Deleted: dc2e1697e33eb13cc6fae03ab3e2c923ac211863a8a2ad1d90a4969bbb5034c1
Deleted: df2a0347c9d081fa05ecb83669dcae5830c67b0676a6d6358218e55d8a45969c
Deleted: 39bb80489af75406073b5364c9c326134015140e1f7976a370a8bd446889e6f8

**Comment by thaJeztah** [comment]
ping @jfrazelle I was able to reproduce this on 1.7-rc1 with the steps above, but it's late here, so ..


```
CONTAINER ID        IMAGE                        COMMAND                CREATED             STATUS              PORTS                                           NAMES
1bd71ba53113        nginx                        "nginx -g 'daemon of   6 minutes ago       Up 6 minutes        0.0.0.0:32780->80/tcp, 0.0.0.0:32779->443/tcp   gloomy_stallman            

docker inspect -f "{{ .Image }}" 1bd71ba53113
a785ba7493fd9009be71023045790697f334486b11a45f7e830652d39665a451

docker inspect a785ba7493fd9009be71023045790697f334486b11a45f7e830652d39665a451
Error: No such image or container: a785ba7493fd9009be71023045790697f334486b11a45f7e830652d39665a451
[]
```

**Comment by jfrazelle** [comment]
super confused because there is a unit test for that https://github.com/docker/docker/blob/master/integration-cli/docker_cli_rmi_test.go#L105

**Comment by duglin** [comment]
I wonder if this is how I got into the bad situation for https://github.com/docker/docker/pull/13527

**Comment by thaJeztah** [comment]
@duglin Not sure, because the "Image" field of the container is not cleared in this case. I was also able to remove the container afterwards without problems.

To verify it wasn't just something with my host, I just spun up a fresh "Debian Jessy" droplet and a "Ubuntu 14.04.2" droplet on DigitalOcean, and was able to reproduce on both.

Note that this *only* occurs if the image has multiple tags locally. If only `nginx:latest` is pulled / present, I get the right error message;

    docker rmi a785ba7493fd9009be71023045790697f334486b11a45f7e830652d39665a451
    Error response from daemon: Conflict, cannot delete a785ba7493fd because the running container a6049358afdd is using it, stop it and use -f to force
    Error: failed to remove images: [a785ba7493fd9009be71023045790697f334486b11a45f7e830652d39665a451]

I'll added it to the 1.7 milestone

**Comment by thaJeztah** [comment]
Okay, was playing a bit with this and it looks like the test doesn't reveal the problem because it also uses different *names* for the images, not different *tags*);

This *is* working as expected (images `busybox-test:latest` and `utest:tag1`);

```bash

CONT_ID=$(docker run -d busybox /bin/sh -c mkdir '/busybox-test')
IMG_ID=$(docker commit $CONT_ID busybox-test)
docker rm $CONT_ID

docker tag busybox-test utest:tag1

CONT_ID=$(docker run -d utest:tag1)
docker rmi $IMG_ID

Error response from daemon: Conflict, cannot delete image 8b267471440e594e81b9aab4e5efcad508cebcf572fdfbb7a029c080e85a8463 because it is tagged in multiple repositories, use -f to force
Error: failed to remove images: [8b267471440e594e81b9aab4e5efcad508cebcf572fdfbb7a029c080e85a8463]
```

But this *isn't*  (images `testb:latest` and `testb:tag1`); 

Docker doesn't detect the conflict and removes the image. It also ignores the container using the image.

```bash

CONT_ID=$(docker run -d busybox /bin/sh -c mkdir '/busybox-test')
IMG_ID=$(docker commit $CONT_ID testb)
docker rm $CONT_ID

docker tag testb testb:tag1

CONT_ID=$(docker run -d testb:tag1)
docker rmi $IMG_ID

Untagged: testb:latest
Untagged: testb:tag1
Deleted: ed966d9a8095bb21794cab72d99592327df47a50e78c8125e48294ff981fa52a
```

**Comment by jfrazelle** [comment]
cool I can take a look today, thanks for finding this :)

**Comment by jfrazelle** [comment]
ok made a failing test case, I will have this fixed shortly ;)

**Comment by thaJeztah** [comment]
@jfrazelle thanks! looked a bit at the code earlier, could use some simplification, but didn't spot the bug directly

**Comment by jfrazelle** [comment]
that code desperitely needs to be refactored but for now im just patching, then will refactor after

**Comment by thaJeztah** [comment]
Thanks for reporting the regression, @softboysxp! A new test was added and a fix added in https://github.com/docker/docker/issues/13596 to resolve this issue, which will be included in the 1.7 release.
